File: generate_favicons.py
#!/usr/bin/env python3
"""
Generate favicon files in multiple sizes from the logo.svg file.
Requires: pip install cairosvg pillow
"""

import os
from pathlib import Path

try:
    import cairosvg
    from PIL import Image
    import io
except ImportError:
    print("Error: Required libraries not found.")
    print("Please install them with: pip install cairosvg pillow")
    exit(1)

# Define the sizes needed
FAVICON_SIZES = {
    'favicon-16x16.png': 16,
    'favicon-32x32.png': 32,
    'favicon-48x48.png': 48,
    'apple-touch-icon-152x152.png': 152,
    'apple-touch-icon-167x167.png': 167,
    'apple-touch-icon.png': 180,
}

def generate_favicons():
    """Generate all favicon sizes from logo.svg"""
    # Get the script directory
    script_dir = Path(__file__).parent
    images_dir = script_dir / 'images'

    # Source SVG file
    source_svg = images_dir / 'logo.svg'

    if not source_svg.exists():
        print(f"Error: Source file not found: {source_svg}")
        return

    print(f"Generating favicons from {source_svg}...")

    # Create images directory if it doesn't exist
    images_dir.mkdir(exist_ok=True)

    # Generate each size
    for filename, size in FAVICON_SIZES.items():
        output_path = images_dir / filename

        try:
            # Convert SVG to PNG at the specified size
            png_data = cairosvg.svg2png(
                url=str(source_svg),
                output_width=size,
                output_height=size
            )

            # Open with PIL and save (for optimization)
            img = Image.open(io.BytesIO(png_data))
            img.save(output_path, 'PNG', optimize=True)

            print(f"✓ Created {filename} ({size}x{size})")

        except Exception as e:
            print(f"✗ Failed to create {filename}: {e}")

    print("\nFavicon generation complete!")
    print("\nGenerated files:")
    for filename in FAVICON_SIZES.keys():
        file_path = images_dir / filename
        if file_path.exists():
            size_kb = file_path.stat().st_size / 1024
            print(f"  - {filename} ({size_kb:.2f} KB)")

if __name__ == '__main__':
    generate_favicons()

File: generate_favicons.js
#!/usr/bin/env node
/**
 * Generate favicon files in multiple sizes from the logo.svg file.
 * Requires: npm install sharp
 */

const fs = require('fs');
const path = require('path');

// Check if sharp is installed
let sharp;
try {
    sharp = require('sharp');
} catch (err) {
    console.error('Error: sharp module not found.');
    console.error('Please install it with: npm install sharp');
    process.exit(1);
}

// Define the sizes needed
const FAVICON_SIZES = {
    'favicon-16x16.png': 16,
    'favicon-32x32.png': 32,
    'favicon-48x48.png': 48,
    'apple-touch-icon-152x152.png': 152,
    'apple-touch-icon-167x167.png': 167,
    'apple-touch-icon.png': 180,
};

async function generateFavicons() {
    const imagesDir = path.join(__dirname, 'images');
    const sourceSvg = path.join(imagesDir, 'logo.svg');

    if (!fs.existsSync(sourceSvg)) {
        console.error(`Error: Source file not found: ${sourceSvg}`);
        return;
    }

    console.log(`Generating favicons from ${sourceSvg}...`);

    // Create images directory if it doesn't exist
    if (!fs.existsSync(imagesDir)) {
        fs.mkdirSync(imagesDir, { recursive: true });
    }

    // Generate each size
    for (const [filename, size] of Object.entries(FAVICON_SIZES)) {
        const outputPath = path.join(imagesDir, filename);

        try {
            await sharp(sourceSvg)
                .resize(size, size, {
                    fit: 'contain',
                    background: { r: 0, g: 0, b: 0, alpha: 0 }
                })
                .png({ quality: 100, compressionLevel: 9 })
                .toFile(outputPath);

            const stats = fs.statSync(outputPath);
            const sizeKb = (stats.size / 1024).toFixed(2);
            console.log(`✓ Created ${filename} (${size}x${size}) - ${sizeKb} KB`);

        } catch (err) {
            console.error(`✗ Failed to create ${filename}:`, err.message);
        }
    }

    console.log('\nFavicon generation complete!');
}

generateFavicons().catch(err => {
    console.error('Error generating favicons:', err);
    process.exit(1);
});
